src/developer/core/plugin-api.md

```diff
@@ -69,16 +72,159 @@ This is the workhorse.
+| `on_notify(self, notification)` | Called once for each [`notification`](/developer/core/network-api/#notification-message).                                                                                                                                                                                                                                                                                                                                                            |
+
+Custom data example:
+```python
+from plugin import Plugin
+
+CUSTOM_TOPIC = "custom_topic"
+
+
+class CustomDataExample(Plugin):
+    def recent_events(self, events):
+        custom_datum = {
+            "topic": CUSTOM_TOPIC,
+            "timestamp": self.g_pool.get_timestamp(),  # Timestamp in pupil time
+            "custom field": 42,
+            # Further fields can be added here.
+            # Their values should be serializable with msgpack.
+        }
+        events[CUSTOM_TOPIC] = [custom_datum]
+```
+
+| UI interaction callbacks                    | Description                                                                                                                                                                                                                                                                               |
+|---------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `on_click(self, pos, button, action)`       | Gets called when the user clicks in the window screen and the event has not been consumed by the GUI. Return `True` if the event was consumed and should not be propagated to any other plugin.                                                                                           |
+| `on_pos(self, pos)`                         | Gets called when the user moves the mouse in the window screen.                                                                                                                                                                                                                           |
+| `on_key(self, key, scancode, action, mods)` | Gets called on key events that were not consumed by the GUI.  Return `True` if the event was consumed and should not be propagated to any other plugin.  See http://www.glfw.org/docs/latest/input_guide.html#input_key for more information key events.                                  |
```

Comment:
Shall we have the glfw links as in-text links in keeping with the rest of the docs?

Reply:
```suggestion
| `on_key(self, key, scancode, action, mods)` | Gets called on key events that were not consumed by the GUI.  Return `True` if the event was consumed and should not be propagated to any other plugin.  See [the GLFW documentation](http://www.glfw.org/docs/latest/input_guide.html#input_key) for more information on key events.       |
```